feat: minimum implementation extending kotlin 2.2 to support ohos target

linhandev opened the pull request.

linhandev committed:
feat: efficient channel to call napi

Signed-off-by: linhandev <hanlin21@huawei.com>

linhandev committed:
extend kotlin native 2.2 to support targeting ohos

Signed-off-by: linhandev <hanlin21@huawei.com>

tencent-adm commented:
[![CLA assistant check](https://cla.opensource.tencent.com/pull/badge/signed)](https://cla.opensource.tencent.com/Tencent-TDS/KuiklyBase-kotlin?pullRequest=12) <br/>All committers have signed the CLA.

linhandev committed:
fix: fill in linking options for hilog

Signed-off-by: linhandev <hanlin21@huawei.com>

linhandev committed:
chore: add license headers

Signed-off-by: linhandev <hanlin21@huawei.com>

linhandev committed:
chore: track elapsed time in build script

linhandev committed:
fix: run build kn workflow on current branch

linhandev committed:
fix: install git on linux

linhandev committed:
chore: check env non-mac

linhandev committed:
fix: install -y

bennyhuo commented on scripts/build-ohos.sh:
```diff
@@ -0,0 +1,124 @@
+
+# settings
+DEPLOY_VERSION=${DEPLOY_VERSION:-2.2.255-SNAPSHOT}
+export KONAN_DATA_DIR=${KONAN_DATA_DIR:-$ROOT_DIR/build/cache/konan}
+export GRADLE_USER_HOME=${GRADLE_USER_HOME:-$ROOT_DIR/build/cache/gradle}
```
这两行不是必须得吧。修改 Gradle Home 后，我这边出现了自动下载 JDK 的逻辑失败，具体提现为 JDK 21 无法自动下载。

linhandev replied:
JDK21我这边一直不能自动下载，我们去掉这个试试

加这个主要是打包过程中gradle stop,不指定一个独特的home容易stop掉其他的任务

linhandev replied:
```
* What went wrong:
Unable to download toolchain matching the requirements ({languageVersion=21, vendor=any vendor, implementation=vendor-specific, nativeImageCapable=false}) from 'null', due to: No defined toolchain download url for MAC_OS on aarch64 architecture.

* Try:
> Learn more about toolchain auto-detection and auto-provisioning at https://docs.gradle.org/8.14/userguide/toolchains.html#sec:auto_detection.
> Learn more about toolchain repositories at https://docs.gradle.org/8.14/userguide/toolchains.html#sub:download_repositories.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
➜  KuiklyBase-kotlin git:(kuikly-base/2.2.0) /usr/libexec/java_home -V

Matching Java Virtual Machines (2):
    24.0.2 (arm64) "Eclipse Adoptium" - "OpenJDK 24.0.2" /Library/Java/JavaVirtualMachines/temurin-24.jdk/Contents/Home
    1.8.0_462 (arm64) "Azul Systems, Inc." - "Zulu 8.88.0.19" /Library/Java/JavaVirtualMachines/zulu-8.jdk/Contents/Home
/Library/Java/JavaVirtualMachines/temurin-24.jdk/Contents/Home
➜  KuiklyBase-kotlin git:(kuikly-base/2.2.0) 
```

确认了不是换gradle home导致，jdk 21 gradle现在看起来就是不支持自动下载的，这块目前写了文档指导开发者出宝前通过brew安装。autoprovision功能计划issue #15 单独跟踪

@bennyhuo

bennyhuo replied:
这个还是有些奇怪的，官方不知道是怎么解决的。不过说起来，即便能自动匹配到版本，也很容易因为网络波动出现失败，我在构建环境里面提前安装了 Kotlin 编译需要的所有 JDK 版本（为了兼容性，Kotlin 需要依赖不同的 JDK 版本），以避免出错。

bennyhuo commented on scripts/build-ohos.sh:
```diff
@@ -0,0 +1,124 @@
+
+stepBegin "Bundle Kotlin Native compiler."
+./gradlew --stop
+GRADLE_NATIVE :kotlin-native:bundle
```
鸿蒙和 apple 一起编译，会导致 llvm 混用，有验证 iOS 的系统 API 是否能正常生成吗？

linhandev replied:
iOS上ovComposeSample跑起来正常。2.2上的修改只最低限度的替换llvm，鸿蒙的llvm只在鸿蒙的 .o -> .so 步骤用到了，理论上也不会导致iOS platformlib有问题

bennyhuo replied:
好的，LLVM 的使用逻辑 2.2 官方修改的还挺多。这个先这这样吧，等后面再仔细对照一下。

bennyhuo commented on kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/cexport/CAdapterCompile.kt:
```diff
@@ -16,6 +16,7 @@ fun produceCAdapterBitcode(clang: ClangArgs, cppFile: File, bitcodeFile: File) {
     val clangCommand = clang.clangCXX(
             "-std=c++17",
             cppFile.absoluteFile.normalize().path,
+            "-O3", "-Xclang", "-disable-llvm-optzns",
```
这样会影响所有生成 C API 的目标平台（例如 Android Native、Linux 等等）。建议只针对鸿蒙平台做参数修改，例如在 ClangArgs 中参考 Apple 平台添加参数。

linhandev replied:
调试时尝试的一个选项不小心带上了，已回滚

bennyhuo commented on kotlin-native/platformLibs/src/platform/ohos/posix.def:
```diff
@@ -0,0 +1,80 @@
+headers = alloca.h ar.h assert.h complex.h ctype.h dirent.h dlfcn.h err.h errno.h fcntl.h \
+    fenv.h fnmatch.h ftw.h getopt.h grp.h inttypes.h libgen.h limits.h \
+    locale.h math.h memory.h netdb.h paths.h poll.h \
+    pthread.h pwd.h regex.h resolv.h sched.h search.h semaphore.h setjmp.h signal.h \
```
2.0.21 上如果不加 threading.h，会导致 Windows 上编译失败，不知道 2.2 这里有没有验证 Windows 的编译？

linhandev replied:
有的，windows上出包这里没报过问题

bennyhuo commented on native/utils/src/org/jetbrains/kotlin/konan/target/ClangArgs.kt:
```diff
@@ -172,14 +176,21 @@ sealed class ClangArgs(
-            // TODO: Revise after update to LLVM 10.
-            listOf("-isystem", "$absoluteLlvmHome/lib/clang/${configurables.llvmVersion}/include")
-
+    private val libclangSpecificArgs = if (configurables.target.family == Family.OHOS) {
```
2.0.21 版本上这块儿没有专门对鸿蒙进行处理。建议参考一下最新代码的修改看看是否可以简化。整体上减少对 ohos target 的判断可以尽可能降低后续的维护成本。

这个 commit 在支持Windows、Linux、macos 三端编译鸿蒙时，对鸿蒙 toolchain 的配置也做了一些优化，可以参考下：
e04e136223dc56ab741f0f39fe70835f93997a73

linhandev replied:
使用oh的构建脚本出的llvm结构和apple版本有差异导致需要配这几行，konan.properties中不支持一个配置项中部分是变量替换成世纪值，部分是硬编码的字符串所以用kt代码做的配置

bennyhuo commented on kotlin-native/runtime/src/main/cpp/Porting.cpp:
```diff
@@ -18,6 +18,12 @@
 #endif
+#ifdef KONAN_OHOS
+#include <hilog/log.h>
+// region Tencent Code
```
我们在后续的代码中增加了 // region Tencent Code ... // endregion 来标识出来相对官方的定制修改。建议整体按照这个规范做一下调整。

linhandev replied:
已添加，建议给单行的修改也定一个格式，比如 // Tencent Code 写在行尾，这样注释能少一点

linhandev committed:
feat: tweak compile options

linhandev committed:
nit: spare cache

linhandev committed:
fix: only mv when cache folder exists

bennyhuo commented on native/utils/src/org/jetbrains/kotlin/konan/target/ConfigurablesImpl.kt:
```diff
@@ -28,6 +29,63 @@ class GccConfigurablesImpl(target: KonanTarget, properties: Properties, dependen
+        } else {
+            throw IllegalStateException("Unsupported host: ${HostManager.host}")
+        }
+    }
```
在 2.0.21 版本最新的一笔提交中，适配了 Windows 和 Linux，这个文件有较大的调整，建议参考合入。

linhandev replied:
内部使用过程中一些开发readme看的不够仔细，经常有人因为装错devecostudio版本导致项目出包失败。我这边感觉
- 相比从kotlin仓库出包，需要特定sysroot版本的频率低很多
- 开发过程中一个团队的开发包括出包的流水线上最好sysroot版本能保持一致避免出现版本问题
基于这两个考虑上次提交review之后我们把pick up用户本地sysroot的功能去掉了。不知道 @bennyhuo 是不是赞成这个选择，或者霍老师之前做的时候是为啥加了这个能力。

当前我们切换sysroot版本是通过在konan.properties中设置绝对路径的方式

bennyhuo replied:
我们内部其实也是写死路径的（Kotlin 官方自己其实也使用相同的方式处理苹果的 SDK 路径），但由我们外发会有法律风险（我们无权重新发布鸿蒙 SDK），当然你们自己发是没什么问题的。。。

linhandev committed:
fix: subtract so's base address in ohos stack trace address and fix symbolication

linhandev committed:
feat: print fatal message on ohos containing kt stacktrace addresses

linhandev committed:
nit: remote custom cache dir settings

linhandev committed:
nit: revert compile command

linhandev committed:
fix: fill in ohos enum

Commit:
add region Tencent Code tag

linhandev committed:
Fix mvnw wrapper validation error on macOs 15.5

Backported from https://github.com/apache/maven-wrapper/pull/155/files
since the fixed version was not yet published.

KTI-2387

linhandev committed:
feat: adjust performance test to run on ohos

linhandev committed:
fix: add more kexe with hdc steps

linhandev committed:
fix: strip logging output

linhandev committed:
nit: surround patch with Tencent Code region

linhandev committed:
nit: typo

linhandev committed:
nit: change folder to send kexe

linhandev commented:
上次review后针对review意见做了修改/回复，调通了 kotlin-native/performance 在鸿蒙target上的运行

requesting another review @bennyhuo

linhandev committed:
feat: support asan on ohos target

sunwarm2001 committed:
feat: send test harmony device

bennyhuo commented:
好的，感谢，我再仔细看一下。
